swift_browser_ui/common/vault_client.py

```diff
@@ -0,0 +1,315 @@
+        )
+    reason = f"Vault error: {reason}"
+    if status >= 500:
+        return VaultServerError(text=reason, reason=reason)
+    return VaultClientError(text=reason, reason=reason, status_code=status)
+
+
+def _process_error(vault_error: VaultError) -> str:
+    """Return formatted error message.
+
+    https://developer.hashicorp.com/vault/api-docs#error-response
+
+    :param vault_error: Vault error dictionary object
+    :returns: formatted error message as a string.
+    """
+    return vault_error.errors[0]
+
+
+class VaultClient:
+    """AioHttp client that interacts with a Vault server"""
+
+    _http_client: Optional[ClientSession] = None
+    _vault_token = ""
+    service = "SD-Connect"
+
+    def __init__(self):
+        """ """
+        self._base_url = URL(os.environ.get("VAULT_URL", "http://localhost:8200")) / "v1"
+        self._role = os.environ.get("VAULT_ROLE", "")
+        self._secret = os.environ.get("VAULT_SECRET", "")
+        self._token_expires = 0
+
+    async def _token(self) -> str:
+        if not self._vault_token or (self._token_expires - time.monotonic()) < 10:
+            auth = {
+                "role_id": self._role,
+                "secret_id": self._secret,
+            }
+            LOGGER.debug(
+                "Attempting to create a new token with role: %r, and secret: %r",
+                self._role,
+                self._secret,
+            )
+            client = await self._client
+            async with client.post(
+                self._base_url / "auth/approle/login", json=auth
+            ) as auth_response:
+                response = await auth_response.json()
+                if auth_response.ok:
+                    self._token_expires = (
+                        time.monotonic() + response["auth"]["lease_duration"]
+                    )
+                    self._vault_token = response["auth"]["client_token"]
+                    LOGGER.debug("New token generated")
+
+                else:
+                    message = _process_error(VaultError(**response))
+                    reason = f"Vault client was unable to authenticate: {message}"
+                    LOGGER.error(reason)
+                    raise VaultServerError(reason=reason)
+
+        return self._vault_token
+
+    @property
+    async def _client(self) -> ClientSession:
+        """Singleton http client, customized for use with Vault."""
+        if self._http_client is None or self._http_client.closed:
+            self._http_client = ClientSession(
+                timeout=ClientTimeout(total=10),
+                # headers={
+                #     "X-Vault-Token": await self._token(),
+                #     "X-Vault-Request": "true",
+                #     "Accept": "application/json",
+                #     "Content-Type": "application/json",
+                # },
+            )
+            LOGGER.debug("New HTTP client created")
+        return self._http_client
```

Comment:
Should be possible to use the client saved in `aiohttp.app` when spinning up the server, since the client has a connection pool.